File: cpp/tests/test_stat.cpp
#include <catch2/catch_test_macros.hpp>
#include <vost/vost.h>

#include <filesystem>
#include <string>
#include <thread>
#include <chrono>

namespace fs = std::filesystem;

static fs::path make_temp_repo() {
    auto tmp = fs::temp_directory_path() /
               ("vost_stest_" + std::to_string(
                    std::hash<std::thread::id>{}(std::this_thread::get_id())
                    ^ static_cast<size_t>(
                          std::chrono::steady_clock::now()
                              .time_since_epoch()
                              .count())));
    return tmp;
}

static vost::GitStore open_store(const fs::path& path,
                                  const std::string& branch = "main") {
    vost::OpenOptions opts;
    opts.create = true;
    opts.branch = branch;
    return vost::GitStore::open(path, opts);
}

// ---------------------------------------------------------------------------
// stat on files
// ---------------------------------------------------------------------------

TEST_CASE("stat: regular file", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("hello.txt", "world");

    auto st = snap.stat("hello.txt");
    CHECK(st.file_type == vost::FileType::Blob);
    CHECK(st.mode == vost::MODE_BLOB);
    CHECK(st.size == 5);  // "world"
    CHECK(st.hash.size() == 40);
    CHECK(st.nlink == 1);
    CHECK(st.mtime > 0);

    fs::remove_all(path);
}

TEST_CASE("stat: executable file", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");

    vost::WriteOptions wopts;
    wopts.mode = vost::MODE_BLOB_EXEC;
    snap = snap.write_text("run.sh", "#!/bin/sh", wopts);

    auto st = snap.stat("run.sh");
    CHECK(st.file_type == vost::FileType::Executable);
    CHECK(st.mode == vost::MODE_BLOB_EXEC);
    CHECK(st.nlink == 1);

    fs::remove_all(path);
}

TEST_CASE("stat: symlink", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("target.txt", "data");
    snap = snap.write_symlink("link.txt", "target.txt");

    auto st = snap.stat("link.txt");
    CHECK(st.file_type == vost::FileType::Link);
    CHECK(st.mode == vost::MODE_LINK);
    CHECK(st.nlink == 1);

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// stat on directories
// ---------------------------------------------------------------------------

TEST_CASE("stat: directory", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("dir/file.txt", "data");

    auto st = snap.stat("dir");
    CHECK(st.file_type == vost::FileType::Tree);
    CHECK(st.mode == vost::MODE_TREE);
    CHECK(st.nlink >= 2);  // at least 2 for a directory

    fs::remove_all(path);
}

TEST_CASE("stat: root directory", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    auto st = snap.stat();  // root
    CHECK(st.file_type == vost::FileType::Tree);
    CHECK(st.nlink >= 2);

    fs::remove_all(path);
}

TEST_CASE("stat: nlink counts subdirs", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");

    auto batch = snap.batch();
    batch.write_text("sub1/a.txt", "a");
    batch.write_text("sub2/b.txt", "b");
    batch.write_text("top.txt", "t");
    snap = batch.commit();

    auto st = snap.stat();
    // nlink = 2 + number of subdirectories (2)
    CHECK(st.nlink == 4);

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// stat: error cases
// ---------------------------------------------------------------------------

TEST_CASE("stat: nonexistent path throws", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    REQUIRE_THROWS_AS(snap.stat("nope.txt"), vost::NotFoundError);
    fs::remove_all(path);
}

TEST_CASE("stat: hash is consistent across calls", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    auto st1 = snap.stat("file.txt");
    auto st2 = snap.stat("file.txt");
    CHECK(st1.hash == st2.hash);

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// stat: size matches content length
// ---------------------------------------------------------------------------

TEST_CASE("stat: size matches content length", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    std::string content = "Hello, World! This is a longer string.";
    snap = snap.write_text("file.txt", content);

    auto st = snap.stat("file.txt");
    CHECK(st.size == static_cast<uint64_t>(content.size()));

    fs::remove_all(path);
}

TEST_CASE("stat: symlink size is target path length", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_symlink("link", "target.txt");

    auto st = snap.stat("link");
    CHECK(st.size == static_cast<uint64_t>(std::string("target.txt").size()));

    fs::remove_all(path);
}

TEST_CASE("stat: nlink for leaf directory (no subdirs)", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("dir/a.txt", "a");
    snap = snap.write_text("dir/b.txt", "b");

    auto st = snap.stat("dir");
    // Leaf dir: nlink = 2 (self + parent)
    CHECK(st.nlink == 2);

    fs::remove_all(path);
}

TEST_CASE("stat: mtime consistency across calls", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    auto st1 = snap.stat("file.txt");
    auto st2 = snap.stat("file.txt");
    CHECK(st1.mtime == st2.mtime);

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// listdir
// ---------------------------------------------------------------------------

TEST_CASE("stat: listdir names match ls", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("a.txt", "a");
    snap = snap.write_text("b.txt", "b");
    snap = snap.write_text("dir/c.txt", "c");

    auto ls_names = snap.ls();
    auto listdir_entries = snap.listdir();

    std::vector<std::string> listdir_names;
    for (auto& e : listdir_entries) {
        listdir_names.push_back(e.name);
    }
    std::sort(ls_names.begin(), ls_names.end());
    std::sort(listdir_names.begin(), listdir_names.end());
    CHECK(ls_names == listdir_names);

    fs::remove_all(path);
}

TEST_CASE("stat: listdir returns WalkEntry with correct types", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");
    snap = snap.write_text("dir/inner.txt", "inner");

    auto entries = snap.listdir();
    bool found_file = false;
    bool found_dir = false;
    for (auto& e : entries) {
        if (e.name == "file.txt") {
            CHECK(e.mode == vost::MODE_BLOB);
            found_file = true;
        }
        if (e.name == "dir") {
            CHECK(e.mode == vost::MODE_TREE);
            found_dir = true;
        }
    }
    CHECK(found_file);
    CHECK(found_dir);

    fs::remove_all(path);
}

TEST_CASE("stat: listdir on file throws", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    REQUIRE_THROWS_AS(snap.listdir("file.txt"), vost::NotADirectoryError);

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// tree_hash
// ---------------------------------------------------------------------------

TEST_CASE("stat: tree_hash is 40-char hex", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    auto th = snap.tree_hash();
    REQUIRE(th.has_value());
    CHECK(th->size() == 40);

    fs::remove_all(path);
}

TEST_CASE("stat: tree_hash changes on write", "[stat]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "v1");
    auto hash1 = snap.tree_hash();

    snap = snap.write_text("file.txt", "v2");
    auto hash2 = snap.tree_hash();

    CHECK(hash1 != hash2);

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// read range
// ---------------------------------------------------------------------------

TEST_CASE("stat: read_range with offset only", "[stat][read_range]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "Hello World");

    auto data = snap.read_range("file.txt", 6);
    std::string result(data.begin(), data.end());
    CHECK(result == "World");

    fs::remove_all(path);
}

TEST_CASE("stat: read_range with offset and size", "[stat][read_range]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "abcdefghij");

    auto data = snap.read_range("file.txt", 3, 4);
    std::string result(data.begin(), data.end());
    CHECK(result == "defg");

    fs::remove_all(path);
}

TEST_CASE("stat: read_range beyond end is clamped", "[stat][read_range]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "short");

    auto data = snap.read_range("file.txt", 2, 100);
    std::string result(data.begin(), data.end());
    CHECK(result == "ort");

    fs::remove_all(path);
}

TEST_CASE("stat: read_range zero size returns empty", "[stat][read_range]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    auto data = snap.read_range("file.txt", 0, 0);
    CHECK(data.empty());

    fs::remove_all(path);
}

TEST_CASE("stat: read_range with offset at end returns empty", "[stat][read_range]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "data");

    auto data = snap.read_range("file.txt", 100);
    CHECK(data.empty());

    fs::remove_all(path);
}

// ---------------------------------------------------------------------------
// read_by_hash
// ---------------------------------------------------------------------------

TEST_CASE("stat: read_by_hash roundtrip", "[stat][read_by_hash]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "content");

    auto hash = snap.stat("file.txt").hash;
    auto data = snap.read_by_hash(hash);
    std::string result(data.begin(), data.end());
    CHECK(result == "content");

    fs::remove_all(path);
}

TEST_CASE("stat: read_by_hash with range", "[stat][read_by_hash]") {
    auto path = make_temp_repo();
    auto store = open_store(path);
    auto snap = store.branches().get("main");
    snap = snap.write_text("file.txt", "Hello World");

    auto hash = snap.stat("file.txt").hash;
    auto data = snap.read_by_hash(hash, 6, 5);
    std::string result(data.begin(), data.end());
    CHECK(result == "World");

    fs::remove_all(path);
}

File: rs/tests/test_stat.rs
mod common;

use vost::*;

// ---------------------------------------------------------------------------
// stat — files
// ---------------------------------------------------------------------------

#[test]
fn stat_file() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let st = fs.stat("hello.txt").unwrap();
    assert_eq!(st.mode, MODE_BLOB);
    assert_eq!(st.file_type, FileType::Blob);
    assert_eq!(st.size, 5);
    assert_eq!(st.nlink, 1);
    assert_eq!(st.hash.len(), 40);
    assert!(st.hash.chars().all(|c| c.is_ascii_hexdigit()));
}

#[cfg(unix)]
#[test]
fn stat_executable() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    fs.write("run.sh", b"#!/bin/sh", fs::WriteOptions {
        mode: Some(MODE_BLOB_EXEC),
        ..Default::default()
    })
    .unwrap();
    let fs = store.branches().get("main").unwrap();
    let st = fs.stat("run.sh").unwrap();
    assert_eq!(st.mode, MODE_BLOB_EXEC);
    assert_eq!(st.file_type, FileType::Executable);
    assert_eq!(st.nlink, 1);
}

#[cfg(unix)]
#[test]
fn stat_symlink() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    fs.write_symlink("link", "target.txt", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();
    let st = fs.stat("link").unwrap();
    assert_eq!(st.mode, MODE_LINK);
    assert_eq!(st.file_type, FileType::Link);
    assert_eq!(st.nlink, 1);
}

// ---------------------------------------------------------------------------
// stat — directories
// ---------------------------------------------------------------------------

#[test]
fn stat_directory() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let st = fs.stat("dir").unwrap();
    assert_eq!(st.mode, MODE_TREE);
    assert_eq!(st.file_type, FileType::Tree);
    assert_eq!(st.size, 0);
    // dir has no subdirs, so nlink = 2
    assert_eq!(st.nlink, 2);
}

#[test]
fn stat_root() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let st = fs.stat("").unwrap();
    assert_eq!(st.mode, MODE_TREE);
    assert_eq!(st.file_type, FileType::Tree);
    assert_eq!(st.size, 0);
    // root has one subdir ("dir"), so nlink = 2 + 1 = 3
    assert_eq!(st.nlink, 3);
}

// ---------------------------------------------------------------------------
// stat — nlink with multiple subdirs
// ---------------------------------------------------------------------------

#[test]
fn stat_nlink_counts_subdirs() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    let mut batch = fs.batch(Default::default());
    batch.write("a/x.txt", b"x").unwrap();
    batch.write("b/y.txt", b"y").unwrap();
    batch.write("c/z.txt", b"z").unwrap();
    batch.write("file.txt", b"f").unwrap();
    batch.commit().unwrap();
    let fs = store.branches().get("main").unwrap();

    // root has 3 subdirs (a, b, c) → nlink = 2 + 3 = 5
    let st = fs.stat("").unwrap();
    assert_eq!(st.nlink, 5);
}

// ---------------------------------------------------------------------------
// stat — consistency with size() and object_hash()
// ---------------------------------------------------------------------------

#[test]
fn stat_size_matches_size_method() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let st = fs.stat("hello.txt").unwrap();
    assert_eq!(st.size, fs.size("hello.txt").unwrap());
}

#[test]
fn stat_hash_matches_object_hash() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let st = fs.stat("hello.txt").unwrap();
    assert_eq!(st.hash, fs.object_hash("hello.txt").unwrap());
}

// ---------------------------------------------------------------------------
// stat — mtime
// ---------------------------------------------------------------------------

#[test]
fn stat_mtime_consistent() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let st1 = fs.stat("hello.txt").unwrap();
    let st2 = fs.stat("dir/a.txt").unwrap();
    // Same commit → same mtime
    assert_eq!(st1.mtime, st2.mtime);
    assert!(st1.mtime > 0);
}

// ---------------------------------------------------------------------------
// stat — nonexistent
// ---------------------------------------------------------------------------

#[test]
fn stat_nonexistent_errors() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    assert!(fs.stat("nope.txt").is_err());
}

// ---------------------------------------------------------------------------
// listdir
// ---------------------------------------------------------------------------

#[test]
fn listdir_matches_ls() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let ls_names: Vec<String> = fs.ls("").unwrap();
    let ld_names: Vec<String> = fs.listdir("").unwrap().iter().map(|e| e.name.clone()).collect();
    assert_eq!(ls_names, ld_names);
}

#[test]
fn listdir_subdir() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let entries = fs.listdir("dir").unwrap();
    let names: Vec<&str> = entries.iter().map(|e| e.name.as_str()).collect();
    assert!(names.contains(&"a.txt"));
    assert!(names.contains(&"b.txt"));
    assert_eq!(entries.len(), 2);
}

// ---------------------------------------------------------------------------
// tree_hash
// ---------------------------------------------------------------------------

#[test]
fn tree_hash_is_40_hex() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let hash = fs.tree_hash().unwrap();
    assert_eq!(hash.len(), 40);
    assert!(hash.chars().all(|c| c.is_ascii_hexdigit()));
}

#[test]
fn tree_hash_changes_on_write() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    fs.write("a.txt", b"hello", Default::default()).unwrap();
    let fs1 = store.branches().get("main").unwrap();
    let h1 = fs1.tree_hash().unwrap();

    fs1.write("b.txt", b"world", Default::default()).unwrap();
    let fs2 = store.branches().get("main").unwrap();
    let h2 = fs2.tree_hash().unwrap();

    assert_ne!(h1, h2);
}

// ---------------------------------------------------------------------------
// read_range
// ---------------------------------------------------------------------------

#[test]
fn read_range_offset_and_size() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    fs.write("data.txt", b"hello world", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();

    // Read "world" (offset=6, size=5)
    let data = fs.read_range("data.txt", 6, Some(5)).unwrap();
    assert_eq!(data, b"world");
}

#[test]
fn read_range_offset_only() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    fs.write("data.txt", b"hello world", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();

    // Read from offset=6 to end
    let data = fs.read_range("data.txt", 6, None).unwrap();
    assert_eq!(data, b"world");
}

#[test]
fn read_range_full() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let full = fs.read("hello.txt").unwrap();
    let ranged = fs.read_range("hello.txt", 0, None).unwrap();
    assert_eq!(full, ranged);
}

#[test]
fn read_range_size_beyond_end() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    // "hello" is 5 bytes; requesting 100 from offset 2 should clamp
    let data = fs.read_range("hello.txt", 2, Some(100)).unwrap();
    assert_eq!(data, b"llo");
}

#[test]
fn read_range_offset_at_end() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let data = fs.read_range("hello.txt", 5, Some(10)).unwrap();
    assert!(data.is_empty());
}

#[test]
fn read_range_offset_beyond_end() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let data = fs.read_range("hello.txt", 100, None).unwrap();
    assert!(data.is_empty());
}

// ---------------------------------------------------------------------------
// read_by_hash
// ---------------------------------------------------------------------------

#[test]
fn read_by_hash_full() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    let hash = fs.object_hash("hello.txt").unwrap();
    let data = fs.read_by_hash(&hash, 0, None).unwrap();
    assert_eq!(data, b"hello");
}

#[test]
fn read_by_hash_ranged() {
    let dir = tempfile::tempdir().unwrap();
    let store = common::create_store(dir.path(), "main");
    let fs = store.branches().get("main").unwrap();
    fs.write("data.txt", b"abcdefghij", Default::default()).unwrap();
    let fs = store.branches().get("main").unwrap();

    let hash = fs.object_hash("data.txt").unwrap();
    let data = fs.read_by_hash(&hash, 3, Some(4)).unwrap();
    assert_eq!(data, b"defg");
}

#[test]
fn read_by_hash_roundtrip_with_object_hash() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    for path in &["hello.txt", "dir/a.txt", "dir/b.txt"] {
        let hash = fs.object_hash(path).unwrap();
        let by_hash = fs.read_by_hash(&hash, 0, None).unwrap();
        let by_path = fs.read(path).unwrap();
        assert_eq!(by_hash, by_path);
    }
}

#[test]
fn read_by_hash_invalid_hash_errors() {
    let dir = tempfile::tempdir().unwrap();
    let (_, fs) = common::store_with_files(dir.path());
    assert!(fs.read_by_hash("not_a_valid_hash", 0, None).is_err());
}
